Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp

```diff
@@ -32,311 +32,259 @@
 {
-    // Process the leftover data.
+    if (m_state == State::WaitingForTerminate || m_state == State::End)
+        return; // The handler is closed down so ignore everything.
+
+    RELEASE_ASSERT(m_state == State::WaitingForHeaderProcessing);
+
+    m_state = State::InBody;
+
     while (processContent()) { }
+}
 
-    if (m_state != State::End) {
-        // It seems we are still not at the end of the processing.
-        // Push out the remaining data.
-        m_client.didReceiveDataFromMultipart(SharedBuffer::create(WTFMove(m_buffer)));
-        m_state = State::End;
-    }
+void CurlMultipartHandle::didCompleteMessage()
+{
+    if (m_state == State::End)
+        return; // The handler is closed down so ignore everything.
+
+    m_didCompleteMessage = true;
 
-    m_buffer.clear();
+    if (m_state == State::WaitingForTerminate)
+        m_state = State::Terminating;
+
+    while (processContent()) { }
 }
 
 bool CurlMultipartHandle::processContent()
 {
-/*
-    The allowed transitions between the states:
-         Check Boundary
-               |
-      /-- In Boundary <----\
-     |         |            |
-     |     In Header        |
-     |         |            |
-     |     In Content       |
-     |         |            |
-     |    End Boundary ----/
-     |         |
-      \-----> End
-
-*/
     switch (m_state) {
-    case State::CheckBoundary: {
-        if (m_buffer.size() < m_boundary.length()) {
-            // We don't have enough data, so just skip.
+    case State::FindBoundaryStart:
+        FALLTHROUGH;
+    case State::InBody: {
+        auto result = findBoundary();
+        if (result.isSyntaxError) {
+            m_hasError = true;
+            m_state = State::End;
             return false;
         }
 
-        // Check for the boundary string.
-        size_t boundaryStart;
-        size_t lastPartialMatch;
+        if (m_state == State::InBody && result.dataEnd)
+            m_client->didReceiveDataFromMultipart(SharedBuffer::create(m_buffer.data(), result.dataEnd));
 
-        if (!checkForBoundary(boundaryStart, lastPartialMatch) && boundaryStart == notFound) {
-            // Did not find the boundary start in this chunk.
-            // Skip ahead to the last valid looking boundary character and start again.
-            m_buffer.remove(0, lastPartialMatch);
-            return false;
-        }
+        if (result.processed)
+            m_buffer.remove(0, result.processed);
 
-        // Found the boundary start.
-        // Consume everything before that and also the boundary
-        m_buffer.remove(0, boundaryStart + m_boundary.length());
-        m_state = State::InBoundary;
-    }
-    // Fallthrough.
-    case State::InBoundary: {
-        // Now the first two characters should be: \r\n
-        if (m_buffer.size() < 2)
-            return false;
+        if (!result.hasBoundary || result.hasCloseDelimiter) {
+            if (shouldTerminateIfNeeded())
+                return true;
```

Comment:
I think this code ends in an infinite loop if the content is terminated without a boundary. You should have a test case.

Reply:
In shouldTerminateIfNeeded(), `m_state` is set to State::Terminating; Therefore, I don't think it will go into an infinite loop.

https://github.com/WebKit/WebKit/pull/17403/files#diff-9748a6c95db83f4a038a66c392226087d69df3953308a3f96c081618d378e750R187-R190

Reply:
The process flow is difficult to understand, so change it as follows.

```
        if (!result.hasBoundary || result.hasCloseDelimiter) {
            if (shouldTerminateIfNeeded()) {
                m_state = State::Terminating;
                return true;
            }
```

Reply:
For example, LayoutTests\http\tests\multipart\multipart-html.py is terminated without a boundary.

Reply:
Changed.

Reply:
Also, I added `NoEndOfBoundaryAfterCompleted`, `NoCloseDelimiterAfterCompleted` and `CloseDelimiterAfterCompleted` for this test.